sycl/include/CL/sycl/handler.hpp

```diff
@@ -742,23 +750,79 @@ class __SYCL_EXPORT handler {
   void parallel_for_lambda_impl(range<Dims> NumWorkItems,
                                 KernelType KernelFunc) {
     throwIfActionIsCreated();
-    using NameT =
-        typename detail::get_kernel_name_t<KernelName, KernelType>::name;
     using LambdaArgType = sycl::detail::lambda_arg_type<KernelType, item<Dims>>;
+
+    // If 1D kernel argument is an integral type, convert it to sycl::item<1>
     using TransformedArgType =
-        typename detail::conditional_t<std::is_integral<LambdaArgType>::value &&
-                                           Dims == 1,
-                                       item<Dims>, LambdaArgType>;
+        typename std::conditional<std::is_integral<LambdaArgType>::value &&
+                                      Dims == 1,
+                                  item<Dims>, LambdaArgType>::type;
+    using NameT =
+        typename detail::get_kernel_name_t<KernelName, KernelType>::name;
+
+    // A reasonable choice for rounding up the range is 32.
+    constexpr size_t GoodLocalSizeX = 32;
+
+    // Disable the rounding-up optimizations under these conditions:
+    // 1. The env var SYCL_OPT_PFWGS_DISABLE is set
+    // 2. When the string SYCL_OPT_PFWGS_DISABLE is in the kernel name.
+    // 3. The kernel is created and invoked without an integration header entry.
+    // 4. The API "this_item" is used inside the kernel.
```

Comment:
What is the rationale for this?
Also point 3 is not obvious to me...

Reply:
1. Allows disabling this transformation at runtime for all parallel_fors without source code changes.
2. Allows disabling this transformation for a specific parallel_for by changing the source code.
3. Disables this optimization when kernel argument setting is done manually by using set_args.

Reply:
My question was actually for the point "4."

Otherwise I find the point "3." quite clearer.
But is the reason that the kernel is provided by an interoperability method instead and thus there is no way to change the kernel from inside? The fact that set_arg is used does not matter I think.

Reply:
3. It is an implementation choice not to support this optimization for the interoperability method. It could conceivably be done but was felt to be of low value. The optimization is intended for the common case of parallel_for.
4. "this_item" is a free function, i.e. not tied to any specific id or item. When concurrent parallel_fors are executing on a device it is not easy to tell which parallel_for the call is being made from, to return the related item. Use of "this_item" is not considered a common case and it is acceptable to give up the optimization when this_item is used within a kernel.

Reply:
Ah thanks, very useful. Now I can understand what's going on.
Could you please copy-paste your explanations in the comments? :-)
I have the feeling that if you detect `this_item` you can clone & rewrite all the call graph to pass explicitly the real item, which looks like some heavy duty... So "giving up" looks like a lazy starting point... :-)

Reply:
I've added to the comments.